<a href="https://colab.research.google.com/github/L-Oliva/Logistic-Regression-Implementation/blob/main/Logistic_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!kaggle datasets download -d fedesoriano/heart-failure-prediction
import pandas as pd
import numpy as np
df = pd.read_csv('/content/heart-failure-prediction.zip')
df

Dataset URL: https://www.kaggle.com/datasets/fedesoriano/heart-failure-prediction
License(s): ODbL-1.0
  0% 0.00/8.56k [00:00<?, ?B/s]
100% 8.56k/8.56k [00:00<00:00, 11.9MB/s]


Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172              N      0.0       Up             0  
1      156              N      1.0     Flat             1  
2       98              N      0.0       Up             0  
3      108              Y      1.5     Flat             1  
4      122              N      0.0       Up             0  
..     ...            ...      ...      ...           ...  
913    132              N      1.2     Flat             1  
914    141              N      3.4     Flat             1  
915    115              Y      1.2     Flat             1  
916    174              N      0.0     Flat             1  
917    173              N      0.0       Up             0  

[918 rows x 12 columns]

In [2]:
heartDisease = df['HeartDisease']
# Droping data that is not in Numeric form
df = df.drop(['HeartDisease','ST_Slope','ExerciseAngina','RestingECG','Sex','ChestPainType'], axis=1)
df = df.to_numpy()
max = np.array([0,0,0,0,0,0])
# Feature Scaling
for i in range(df.shape[0]):
  if df[i][0] > max[0]:
    max[0] = df[i][0]
  if df[i][1] > max[1]:
    max[1] = df[i][1]
  if df[i][2] > max[2]:
    max[2] = df[i][2]
  if df[i][3] > max[3]:
    max[3] = df[i][3]
  if df[i][4] > max[4]:
    max[4] = df[i][4]
  if df[i][5] > max[5]:
    max[5] = df[i][5]

for i in range(df.shape[0]):
  for j in range(df.shape[1]):
    df[i,j] /= max[j]
df

array([[0.51948052, 0.7       , 0.47927032, 0.        , 0.85148515,
        0.        ],
       [0.63636364, 0.8       , 0.29850746, 0.        , 0.77227723,
        0.16666667],
       [0.48051948, 0.65      , 0.46932007, 0.        , 0.48514851,
        0.        ],
       ...,
       [0.74025974, 0.65      , 0.2172471 , 0.        , 0.56930693,
        0.2       ],
       [0.74025974, 0.65      , 0.39137645, 0.        , 0.86138614,
        0.        ],
       [0.49350649, 0.69      , 0.29021559, 0.        , 0.85643564,
        0.        ]])

In [3]:
# Gradient Decent
def decent(heartDisease,df,w,b):
  m,n = df.shape
  dj_dw = np.zeros((n,))
  dj_db = 0
  for i in range(m):
    z = np.dot(df[i],w)+b
    output = 1/(1+np.exp(-z))
    error = output-heartDisease[i]
    for j in range(n):
      dj_dw[j] += error* df[i,j]
    dj_db += error
  return dj_dw/m,dj_db/m

In [4]:
# Cost Function
def cost(heartDisease,X,w,b,max):
  m = X.shape[0]
  cost = 0
  for i in range(m):
    z = np.dot(X[i],w)+b
    output = 1/(1+np.exp(-z))
    cost += -heartDisease[i]*np.log(output) - (1-heartDisease[i])*np.log(1-output)
  return cost * (1/m)

In [5]:
# Prediction
def predict(df,w,b):
  z = np.dot(df[0],w)+b
  output = 1/(1+np.exp(-z))
  return output

In [7]:
w = np.array([1,1,1,1,1,1])
b = 1
a = 1
for x in range(100000):
  dj_dw,dj_db = decent(heartDisease,df,w,b)
  w = w-a*dj_dw
  b = b-a*dj_db
  print(x,"W:",w,"B:",b,"Cost:",cost(heartDisease,df,w,b,max),"Prerdict",predict(df,w,b),"Ans:",heartDisease[0] )


Streaming output truncated to the last 5000 lines.
95000 W: [ 0.79362641  0.46866628 -2.27166779  1.22484761 -5.92127335  5.82624594] B: 3.1312745886180493 Cost: 0.4858244207817356 Prerdict 0.4332285382364521 Ans: 1
95001 W: [ 0.79362641  0.46866628 -2.27166779  1.22484761 -5.92127335  5.82624594] B: 3.1312745886180493 Cost: 0.4858244207817356 Prerdict 0.4332285382364521 Ans: 1
95002 W: [ 0.79362641  0.46866628 -2.27166779  1.22484761 -5.92127335  5.82624594] B: 3.1312745886180493 Cost: 0.4858244207817356 Prerdict 0.4332285382364521 Ans: 1
95003 W: [ 0.79362641  0.46866628 -2.27166779  1.22484761 -5.92127335  5.82624594] B: 3.1312745886180493 Cost: 0.4858244207817356 Prerdict 0.4332285382364521 Ans: 1
95004 W: [ 0.79362641  0.46866628 -2.27166779  1.22484761 -5.92127335  5.82624594] B: 3.1312745886180493 Cost: 0.4858244207817356 Prerdict 0.4332285382364521 Ans: 1
95005 W: [ 0.79362641  0.46866628 -2.27166779  1.22484761 -5.92127335  5.82624594] B: 3.1312745886180493 Cost: 0.48582442078

After 100,000 occurences at a Learning Rate of 0.01
---
99999 W: [ 1.7258804   0.9739464  -2.30522529  1.17714566 -4.72209337  5.6299533 ] B: 1.3744817348858676 Cost: 0.48783899986384405 Prerdict 0.10221859887154464 Ans: 0

After 100,000 occurences at Learning Rate of 0.1
---
99999 W: [ 0.7955132   0.47128912 -2.27185448  1.22471827 -5.91872233  5.82589787] B: 3.126617425061509 Cost: 0.4858244334768553 Prerdict 0.4333508593001198 Ans: 1

After 100,000 occurences at Learning Rate of 1
---
99999 W: [ 0.79362641  0.46866628 -2.27166779  1.22484761 -5.92127335  5.82624594] B: 3.1312745886180493 Cost: 0.4858244207817356 Prerdict 0.4332285382364521 Ans: 1